crates/hyperqueue/src/dashboard/data_source.rs

```diff
@@ -0,0 +1,58 @@
+use crate::dashboard::ui::screen::ClusterState;
+use crate::rpc_call;
+use crate::transfer::connection::ClientConnection;
+use crate::transfer::messages::{FromClientMessage, ToClientMessage};
+use tako::messages::gateway::{CollectedOverview, MonitoringEventRequest};
+use tako::server::monitoring::{MonitoringEvent, MonitoringEventPayload};
+
+async fn get_events_after(
+    connection: &mut ClientConnection,
+    after_id: Option<u32>,
+) -> crate::Result<Vec<MonitoringEvent>> {
+    let response = rpc_call!(
+        connection,
+        FromClientMessage::MonitoringEvents(MonitoringEventRequest {
+            after_id
+        }),
+        ToClientMessage::MonitoringEventsResponse(response) => response
+    )
+    .await;
+    response
+}
+
+#[derive(Default)]
+pub struct DashboardData {
+    /// The event_id until which the data has already been fetched for
+    fetched_until: u32,
+    /// All events received from the client
+    events: Vec<MonitoringEvent>,
+}
+
+impl DashboardData {
+    pub fn get_cluster_state(&self) -> Option<ClusterState> {
+        // let last_overview: CollectedOverview = self
+        //     .events
+        //     .into_iter()
+        //     .filter_map(|event| match event.message {
+        //         MonitoringEventPayload::OverviewUpdate(overview) => Some(overview),
+        //         _ => None,
+        //     })
+        //     .collect::<Vec<CollectedOverview>>()
+        //     .last()
+        //     .unwrap()
+        //     .clone();
+
+        None
+    }
+
+    pub async fn fetch_latest(&mut self, conn: &mut ClientConnection) -> anyhow::Result<()> {
+        let mut new_events = get_events_after(conn, Some(self.fetched_until)).await?;
+        self.events.append(&mut new_events);
```

Comment:
The `append` call will reset `new_events`, so after it the vector will be empty. Also, if there are no new events, you shouldn't reset `fetched_until` to zero.

```rust
self.fetched_until = new_events
    .iter()
    .map(|event| event.id)
    .max()
    .unwrap_or(self.fetched_until);
self.events.append(&mut new_events);
```